### PR TITLE
executor: support SQL mode NO_AUTO_VALUE_ON_ZERO

### PR DESCRIPTION
**Pull request by breeswish**
fix https://github.com/pingcap/tidb/issues/3639

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @XuHuaiyu @winoros @lamxTyler

**Comment by zimulala** [comment]
PTAL @zz-jason

**Comment by zimulala** [comment]
In MySQL version: 5.7.15 the behavior is followings:
```
mysql> show create table t;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` int(11) NOT NULL AUTO_INCREMENT,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`)
) ENGINE=InnoDB AUTO_INCREMENT=11 DEFAULT CHARSET=utf8 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
mysql> insert into t value(0,0);
Query OK, 1 row affected (0.01 sec)
mysql> insert into t value(0,0);
Query OK, 1 row affected (0.01 sec)
mysql> set sql_mode=NO_AUTO_VALUE_ON_ZERO;
Query OK, 0 rows affected, 1 warning (0.00 sec)
mysql> insert into t value(0,0);
ERROR 1062 (23000): Duplicate entry '0' for key 'PRIMARY'
```
You can try on MySQL 5.6.

File: executor/write.go [hunk 1]
```diff
@@ -912,7 +912,7 @@ func (e *InsertValues) initDefaultValues(row []types.Datum, marked map[int]struc
 				return errors.Trace(err)
 			}
 			row[i].SetInt64(val)
-			if val != 0 {
+			if val != 0 || (e.ctx.GetSessionVars().SQLMode&mysql.ModeNoAutoValueOnZero) > 0 {
```
Comment (zimulala) [inline executor/write.go]: In MySQL version: 5.7.15 the behavior is followings:
```
mysql> show create table t;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                     |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` int(11) NOT NULL AUTO_INCREMENT,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`)
) ENGINE=InnoDB AUTO_INCREMENT=11 DEFAULT CHARSET=utf8 |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
mysql> insert into t value(0,0);
Query OK, 1 row affected (0.01 sec)
mysql> insert into t value(0,0);
Query OK, 1 row affected (0.01 sec)
mysql> set sql_mode=NO_AUTO_VALUE_ON_ZERO;
Query OK, 0 rows affected, 1 warning (0.00 sec)
mysql> insert into t value(0,0);
ERROR 1062 (23000): Duplicate entry '0' for key 'PRIMARY'
```
You can try on MySQL 5.6.

Comment (breeswish) [inline executor/write.go]: Strange. My MySQL version is 5.7.18:

```
mysql> drop table t;
Query OK, 0 rows affected (0.08 sec)

mysql> CREATE TABLE `t` (
    ->   `a` int(11) NOT NULL AUTO_INCREMENT,
    ->   `b` int(11) DEFAULT NULL,
    ->   PRIMARY KEY (`a`)
    -> ) ENGINE=InnoDB AUTO_INCREMENT=11 DEFAULT CHARSET=utf8
    -> ;
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t value(0,0);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t value(0,0);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+----+------+
| a  | b    |
+----+------+
| 11 |    0 |
| 12 |    0 |
+----+------+
2 rows in set (0.00 sec)

mysql> set sql_mode=NO_AUTO_VALUE_ON_ZERO;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> insert into t value(0,0);
Query OK, 1 row affected (0.00 sec)
```

Comment (hanfei1991) [inline executor/write.go]: It seems we need a discussion to decide the behaviors.

Comment (zimulala) [inline executor/write.go]: You should do `insert into t value(0,0);` again.

Comment (breeswish) [inline executor/write.go]: If I understood correctly, this scenario is already covered in the test case https://github.com/pingcap/tidb/pull/3661/files#diff-27e6281ea7526ea83fff97f6883294b4R288

Comment (zimulala) [inline executor/write.go]: I so sorry. I made a mistake.

**Pull request merged**